azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -145,30 +143,38 @@ def run(self, goal_state, ext_gs_updated):
             if not conf.get_autoupdate_enabled() or not conf.get_download_new_agents():
                 return
 
-            # verify if agent update is allowed this time (RSM checks new goal state; self-update checks manifest download interval)
-            if not self._updater.is_update_allowed_this_time(ext_gs_updated):
-                return
+            # Update the state only on new goal state
+            if ext_gs_updated:
+                self._gs_id = goal_state.extensions_goal_state.id
+                self._updater.sync_new_gs_id(self._gs_id)
 
-            self._gs_id = goal_state.extensions_goal_state.id
             agent_family = self._get_agent_family_manifest(goal_state)
```

Comment:
I still decide to read properties from goal state on every iteration as goal state is in memory and don't see a big deal. 

Having to call only on new goal state will need to save the state of agent family and also need to maintain state of last attempt has agent family missing error, so that we don't procced with update. I don't like the idea of saving extra state, at some time it will become convoluted.

So I raise different exception and handling it in exception block.